### ISSUE TITLE
Per-service resource limits

### ISSUE THREAD

**Issue by FedericoPonzi** [FedericoPonzi/Horust#12]
It would be nice to have another section for allowing per-service resource limits. This is just a draft and would need some more thoughts on it.

**Comment by kemingy** [comment]
This is a handy feature for long-running services. Is there a plan for what kind of tools to use? As far as I know, both `cgroups` and `eBPF` can do this.

**Comment by FedericoPonzi** [comment]
I think cgroups would be the simplest way to do it, I have some experience with ebpf but decided to ask an llm about pros and cons, let me know what you think

### **eBPF vs cgroups: Pros and Cons**

#### **eBPF**
**Pros:**
1. **Flexibility**: Highly customizable for monitoring and enforcing complex policies.
2. **Kernel-Level Control**: Runs directly in the kernel, enabling low-overhead, real-time actions.
3. **Observability**: Combines resource limiting with deep observability (e.g., tracing, profiling).
4. **Dynamic**: Programs can be loaded/unloaded at runtime without kernel reboots.
5. **Cross-Subsystem**: Can interact with multiple kernel subsystems (networking, scheduling, etc.).

**Cons:**
1. **Complexity**: Requires deep knowledge of kernel internals and eBPF programming.
2. **Indirect Control**: Does not directly limit resources; relies on interactions with other subsystems (e.g., cgroups).
3. **Performance Overhead**: While low, eBPF programs still introduce some overhead.
4. **Limited to Monitoring/Enforcement**: Not a native resource management tool like cgroups.


#### **cgroups**
**Pros:**
1. **Simplicity**: Easy to use for basic resource limiting (CPU, memory, I/O, etc.).
2. **Native Support**: Built into the Linux kernel specifically for resource management.
3. **Fine-Grained Control**: Provides precise control over resource allocation.
4. **Wide Adoption**: Standard tool for container runtimes (e.g., Docker, Kubernetes).
5. **No Programming Required**: Managed via filesystem or tools like `systemd`.

**Cons:**
1. **Static Configuration**: Less flexible for dynamic or custom policies.
2. **Limited Observability**: Does not provide deep insights into resource usage.
3. **No Cross-Subsystem Integration**: Focused solely on resource management.
4. **Requires Root Privileges**: Managing cgroups typically needs elevated permissions.

### **When to Use Which**
- **Use eBPF** if you need advanced, dynamic, or cross-subsystem resource management with observability.
- **Use cgroups** for straightforward, static resource limiting with minimal complexity.

**Comment by FedericoPonzi** [comment]
We can use https://crates.io/crates/cgroups-rs

**Comment by kemingy** [comment]
I agree cgroups is a good fit. I also learned some eBPF but have no scenario to try it
